packages/core/src/main/util/encryptedStore.ts

```diff
@@ -0,0 +1,135 @@
+
+type EncryptedStoreConfig<EncryptedPayload> = {
+  encrypt: EncryptFn<EncryptedPayload>;
+  decrypt: DecryptFn<EncryptedPayload>;
+};
+
+class EncryptedStore<EncryptedPayload> {
+  readonly #decrypt: DecryptFn<EncryptedPayload>;
+  readonly #encrypt: EncryptFn<EncryptedPayload>;
+  constructor(
+    private readonly db: IDBPDatabase<EncryptedDB<EncryptedPayload>>,
+    {encrypt, decrypt}: EncryptedStoreConfig<EncryptedPayload>,
+  ) {
+    this.#encrypt = encrypt;
+    this.#decrypt = decrypt;
+  }
+
+  async saveSecretValue(primaryKey: string, value: Uint8Array) {
+    const encrypted = await this.#encrypt(value);
+    await this.db.put('secrets', encrypted, primaryKey);
+  }
+
+  async getsecretValue(primaryKey: string) {
+    const result = await this.db.get('secrets', primaryKey);
+    if (!result) {
+      return undefined;
+    }
+    return this.#decrypt(result);
+  }
+}
+
+async function generateKey() {
+  return crypto.subtle.generateKey(
+    {name: 'AES-GCM', length: 256},
+    false, //whether the key is extractable (i.e. can be used in exportKey)
+    ['encrypt', 'decrypt'],
+  );
+}
+
+async function defaultDecrypt({value, iv}: DefaultEncryptedPayload, key: CryptoKey): Promise<Uint8Array> {
+  const decrypted = await crypto.subtle.decrypt({name: 'AES-GCM', iv}, key, value);
+  return new Uint8Array(decrypted);
+}
+
+async function defaultEncrypt(data: Uint8Array, key: CryptoKey): Promise<DefaultEncryptedPayload> {
+  const iv = await crypto.getRandomValues(new Uint8Array(12));
+  return {
+    iv,
+    value: await crypto.subtle.encrypt({name: 'AES-GCM', iv}, key, data),
+  };
+}
+
+/**
+ * Will create a database that uses the built in encryption/decryption functions.
+ * The master key will be created and stored inside the database
+ *
+ * @param dbName the name of the database to create
+ */
+export async function createEncryptedStore(dbName: string): Promise<EncryptedStore<DefaultEncryptedPayload>> {
+  const db = await openDB<EncryptedDB<DefaultEncryptedPayload>>(dbName, 1, {
+    upgrade: async database => {
+      database.createObjectStore('key');
+      database.createObjectStore('secrets');
+    },
+  });
+
+  const keyPrimaryKey = 'dbKey';
+  let key = await db.get('key', keyPrimaryKey);
+  if (!key) {
+    key = await generateKey();
+    await db.put('key', key, keyPrimaryKey);
+  }
+  return new EncryptedStore(db, {
+    encrypt: value => defaultEncrypt(value, key as CryptoKey),
+    decrypt: payload => defaultDecrypt(payload, key as CryptoKey),
+  });
+}
+
+/**
+ * Will create a database that uses a custom encryption method. It needs the encrypt and decrypt function to be able to process the values stored.
+ * It's the responsability of the consumer to store the encryption key
+ *
+ * @param dbName the name of the database to create
+ * @param config contains the encrypt and decrypt methods
+ */
+export async function createCustomEncryptedStore<EncryptedPayload>(
```

Comment:
can you explain the use case of custom encryption?

Reply:
If the consumer want to use their own encryption method, the lib allows them to do so. 

A concrete example of this is the electron wrapper. 
Electron has a native way of encrypting secrets (which is more secure that the browser's way). 
So on electron, we will have a custom encryption method that will be given to the core (see https://github.com/wireapp/wire-desktop/pull/5812/files for context) and the core can avoid using the less secure way of generating secrets and rely on electron's system :)